darktable.1
.\" Automatically generated by Pod::Man 2.22 (Pod::Simple 3.07)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.ie \nF \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    nr % 0
.    rr F
.\}
.el \{\
.    de IX
..
.\}
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "DARKTABLE 1"
.TH DARKTABLE 1 "17219d" "darktable-a2ea27c" "darktable"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
darktable \- a digital photography workflow application
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
\&\f(CW\*(C`darktable [options] [IMG_1234.RAW]\*(C'\fR
.PP
Options:
.PP
.Vb 3
\&    \-\-help        
\&    \-\-version
\&    \-d {cache,control,dev,fswatch,perf}
.Ve
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
\&\fBdarktable\fR is a digital photography workflow application for \fBLinux\fR 
and \fBMac \s-1OS\s0 X\fR in the lines of \fBAdobe Lightroom\fR and \fBApple Aperture\fR.
.PP
The application is designed to ease editing and consistent processing of large 
photo sessions and provides a easy to use digital light-table and a set 
of sophisticated postprocessing tools.
.PP
Most processing is done in 32\-bit floating point per channel mode in device independent 
\&\fB\s-1CIE\s0 L*a*b*\fR color space. \fBdarktable\fR is also fully color managed, which gives 
you full control over the look of the photos.
.PP
The application relies on a modern plug-in architecture thus making it easy for 
3rd party developers to extend the existing capabilities of the application. 
All light-table and darkroom features are implemented as plug-ins, so you can 
create your plug-ins reusing existing code.
.SH "COMMAND LINE ARGUMENTS"
.IX Header "COMMAND LINE ARGUMENTS"
.IP "\fB\-d control\fR" 4
.IX Item "-d control"
Enable job queue debugging. If you redirect darktables output to \fBcontrol.log\fR and
call \fB./tools/create_control_svg.sh control.log\fR, you will get a nice \fBcontrol.svg\fR
with a visualization of the threads' work.
.IP "\fB\-d cache\fR" 4
.IX Item "-d cache"
This will give you a lot of debugging info about the mipmap cache for light table mode. If
compiled in debug mode, this will also tell you where in the code a certain buffer has
last been locked.
.IP "\fB\-d perf\fR" 4
.IX Item "-d perf"
Use this for performance tweaking your darkroom modules. It will rdtsc-measure the
runtimes of all plugins and print them to stdout.
.SH "OTHER INFO"
.IX Header "OTHER INFO"
Please visit \fBdarktable\fR's website for news, bug tracker and forum: <http://darktable.sourceforge.net>
.SH "KEYBINDINGS"
.IX Header "KEYBINDINGS"
All modes:
.IP "\fBF11\fR" 4
.IX Item "F11"
Switch between fullscreen and normal modes of the application's window
.IP "\fBTab\fR" 4
.IX Item "Tab"
Show/hide sidebars
.IP "\fBe\fR" 4
.IX Item "e"
Switch between light-table and darkroom modes
.IP "\fBF7\fR" 4
.IX Item "F7"
Decrease contrast
.IP "\fBF8\fR" 4
.IX Item "F8"
Increase contrast
.PP
Lighttable mode:
.IP "\fBalt\-1\fR" 4
.IX Item "alt-1"
Zoom in on image under cursor
.IP "\fBalt\-2, 3\fR" 4
.IX Item "alt-2, 3"
Adjust zoom
.IP "\fBalt\-4\fR" 4
.IX Item "alt-4"
Zoom out completely
.IP "\fBq\fR" 4
.IX Item "q"
Zoom out to fit all images in one row inside viewport, center on image under cursor
.IP "\fBa, s, d, w\fR" 4
.IX Item "a, s, d, w"
Move around (arrows also work)
.IP "\fB1, 2, 3, 4\fR" 4
.IX Item "1, 2, 3, 4"
Star rating
.IP "\fBctrl-backspace\fR" 4
.IX Item "ctrl-backspace"
Strip all stars (reject image)
.IP "\fBF1, F2, F3\fR" 4
.IX Item "F1, F2, F3"
Color labels: toggle red, yellow, green
.IP "\fBctrl-g, ctrl-shift-g\fR" 4
.IX Item "ctrl-g, ctrl-shift-g"
Move to top, bottom, respectively
.IP "\fBctrl-e\fR" 4
.IX Item "ctrl-e"
Export selected images
.PP
Darkroom mode:
.IP "\fBalt\-1, 2, 3\fR" 4
.IX Item "alt-1, 2, 3"
Zoom to 1:1, fill, and fit, respectively
.IP "\fBctrl-f\fR" 4
.IX Item "ctrl-f"
show/hide film strip
.IP "while cropping: \fBEnter/Backspace\fR" 4
.IX Item "while cropping: Enter/Backspace"
apply/revert currently selected cropping frame
.SH "BUGS"
.IX Header "BUGS"
Please use the mailing list to report all the bugs you run into.
.SH "AUTHORS"
.IX Header "AUTHORS"
The principal developer of darktable is Johannes Hanika. The (hopefully)
complete list of contributors to the project is:
.PP
* developers:
Henrik Andersson,
Johannes Hanika.

* ubuntu packaging, color management, video tutorials:
Pascal de Bruijn.

* translation:
Alexandre Prokoudine,
Daniele Giorgis,
Ger Siemerink,
Henrik Andersson,
Josep Puigdemont,
Michał Prędotka,
Mikko Ruohola,
Olivier Tribout,
Roberto Quintero,
Sebastien Delcoigne,
Thomas Costis,
Tobias Ellinghaus.


* contributors:
Alexandre Prokoudine,
Alexander Rabtchevich,
Andrea Purracchio,
Andrey Kaminsky,
Anton Blanchard,
Bernhard Schneider,
Christian Fuchs,
Christian Himpel,
David Bremner,
Gregor Quade,
Jan Rinze,
Jose Carlos Garcia Sogo,
Mikko Ruohola,
Nao Nakashima,
Olivier Tribout,
Pascal de Bruijn,
Robert Park,
Richard Hughes,
Stephen van den Berg,
Stuart Henderson,
Thierry Leconte,
Tobias Ellinghaus,
Wyatt Olson,
Xavier Besse,
Zeus Panchenko.

.PP
This man page was written by Alexandre Prokoudine 
<alexandre.prokoudine@gmail.com>.
.SH "HISTORY"
.IX Header "HISTORY"
The project was started by Johannes Hanika in early 2009 to fill the gap 
(or, rather, a black hole) of a digital photography workflow tool on Linux.
.SH "COPYRIGHT AND LICENSE"
.IX Header "COPYRIGHT AND LICENSE"
\&\fBCopyright (C)\fR 2009\-2010 by Authors.
.PP
\&\fBdarktable\fR is free software; you can redistribute it and/or modify it
under the terms of the \s-1GPL\s0 v3 or (at your option) any later version.


NEWS
see webpage: http://darktable.sf.net/


src/common/darktable.c
/*
    This file is part of darktable,
    copyright (c) 2009--2010 johannes hanika.

    darktable is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    darktable is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with darktable.  If not, see <http://www.gnu.org/licenses/>.
*/
#ifdef HAVE_CONFIG_H
  #include "../config.h"
#endif
#include "common/darktable.h"
#include "common/collection.h"
#include "common/exif.h"
#include "common/fswatch.h"
#include "common/pwstorage/pwstorage.h"
#include "common/camera_control.h"
#include "common/film.h"
#include "common/image.h"
#include "common/image_cache.h"
#include "common/imageio_module.h"
#include "common/points.h"
#include "libs/lib.h"
#include "views/view.h"
#include "control/control.h"
#include "control/conf.h"
#include "gui/gtk.h"
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <glib.h>
#include <lcms.h>
#include <string.h>
#ifdef _OPENMP
#  include <omp.h>
#endif

darktable_t darktable;
const char dt_supported_extensions[] = "3fr,arw,bay,bmq,cap,cine,cr2,crw,cs1,dc2,dcr,dng,erf,fff,hdr,ia,iiq,jpg,jpeg,k25,kc2,kdc,mdc,mef,mos,mrw,nef,nrw,orf,pef,pfm,pxn,qtk,raf,raw,rdc,rw2,rwl,sr2,srf,sti,tif,tiff,x3f";

int dt_init(int argc, char *argv[])
{
  bindtextdomain (GETTEXT_PACKAGE, DARKTABLE_LOCALEDIR);
  bind_textdomain_codeset (GETTEXT_PACKAGE, "UTF-8");
  textdomain (GETTEXT_PACKAGE);

  darktable.progname = argv[0];
#ifdef _OPENMP
  omp_set_num_threads(omp_get_num_procs());
#endif
  darktable.unmuted = 0;
  char *image_to_load = NULL;
  for(int k=1;k<argc;k++)
  {
    if(argv[k][0] == '-')
    {
      if(!strcmp(argv[k], "--help"))
      {
        printf("usage: %s [-d {all,cache,control,dev,fswatch,camctl,pwstorage}] [IMG_1234.{RAW,..}]\n", argv[0]);
        return 1;
      }
      else if(!strcmp(argv[k], "--version"))
      {
        printf("this is "PACKAGE_STRING"\ncopyright (c) 2009-2010 johannes hanika\n"PACKAGE_BUGREPORT"\n");
        return 1;
      }
      if(argv[k][1] == 'd' && argc > k+1)
      {
        if(!strcmp(argv[k+1], "all"))   darktable.unmuted = 0xffffffff;   // enable all debug information
        if(!strcmp(argv[k+1], "cache"))   darktable.unmuted |= DT_DEBUG_CACHE;   // enable debugging for lib/film/cache module
        if(!strcmp(argv[k+1], "control")) darktable.unmuted |= DT_DEBUG_CONTROL; // enable debugging for scheduler module
        if(!strcmp(argv[k+1], "dev"))     darktable.unmuted |= DT_DEBUG_DEV; // develop module
        if(!strcmp(argv[k+1], "fswatch")) darktable.unmuted |= DT_DEBUG_FSWATCH; // fswatch module
        if(!strcmp(argv[k+1], "camctl")) darktable.unmuted |= DT_DEBUG_CAMCTL; // camera control module
        if(!strcmp(argv[k+1], "perf"))    darktable.unmuted |= DT_DEBUG_PERF; // performance measurements
        if(!strcmp(argv[k+1], "pwstorage")) darktable.unmuted |= DT_DEBUG_PWSTORAGE; // pwstorage module
        k ++;
      }
    }
    else
    {
      image_to_load = argv[k];
    }
  }

#ifdef HAVE_GEGL
  (void)setenv("GEGL_PATH", DATADIR"/gegl:/usr/lib/gegl-0.0", 1);
  gegl_init(&argc, &argv);
#endif
  // thread-safe init:
  dt_exif_init();
  (void)cmsErrorAction(LCMS_ERROR_IGNORE);
  char datadir[1024];
  dt_get_user_config_dir (datadir,1024);
  char filename[1024];
  snprintf(filename, 1024, "%s/darktablerc", datadir);

  // Initialize the filesystem watcher  
  darktable.fswatch=dt_fswatch_new();	
  
  // Initialize the camera control 
  darktable.camctl=dt_camctl_new();
  
  // has to go first for settings needed by all the others.
  darktable.conf = (dt_conf_t *)malloc(sizeof(dt_conf_t));
  dt_conf_init(darktable.conf, filename);

  // initialize collection query
  darktable.collection = dt_collection_new(NULL);  

  // Initialize the password storage engine
  darktable.pwstorage=dt_pwstorage_new();	

  char dbfilename[1024];
  gchar *dbname = dt_conf_get_string ("database");
  if(!dbname)               snprintf(dbfilename, 1024, "%s/library.db", datadir);
  else if(dbname[0] != '/') snprintf(dbfilename, 1024, "%s/%s", datadir, dbname);
  else                      snprintf(dbfilename, 1024, "%s", dbname);

  int load_cached = 1;
  // if db file does not exist, also don't load the cache.
  if(!g_file_test(dbfilename, G_FILE_TEST_IS_REGULAR)) load_cached = 0;
  if(sqlite3_open(dbfilename, &(darktable.db)))
  {
    fprintf(stderr, "[init] could not open database ");
    if(dbname) fprintf(stderr, "`%s'!\n", dbname);
    else       fprintf(stderr, "\n");
#ifndef HAVE_GCONF
    fprintf(stderr, "[init] maybe your %s/darktablerc is corrupt?\n",datadir);
    dt_get_datadir(dbfilename, 512);
    fprintf(stderr, "[init] try `cp %s/darktablerc %s/darktablerc'\n", dbfilename,datadir);
#else
    fprintf(stderr, "[init] check your /apps/darktable/database gconf entry!\n");
#endif
    sqlite3_close(darktable.db);
    g_free(dbname);
    return 1;
  }
  g_free(dbname);
  pthread_mutex_init(&(darktable.db_insert), NULL);
  pthread_mutex_init(&(darktable.plugin_threadsafe), NULL);

  darktable.points = (dt_points_t *)malloc(sizeof(dt_points_t));
  dt_points_init(darktable.points, dt_get_num_threads());

  darktable.control = (dt_control_t *)malloc(sizeof(dt_control_t));
  dt_control_init(darktable.control);

  int thumbnails = dt_conf_get_int ("mipmap_cache_thumbnails");
  thumbnails = MIN(1000, MAX(20, thumbnails));

  darktable.mipmap_cache = (dt_mipmap_cache_t *)malloc(sizeof(dt_mipmap_cache_t));
  dt_mipmap_cache_init(darktable.mipmap_cache, thumbnails);

  darktable.image_cache = (dt_image_cache_t *)malloc(sizeof(dt_image_cache_t));
  dt_image_cache_init(darktable.image_cache, MIN(10000, MAX(500, thumbnails)), load_cached);

  darktable.lib = (dt_lib_t *)malloc(sizeof(dt_lib_t));
  dt_lib_init(darktable.lib);

  darktable.view_manager = (dt_view_manager_t *)malloc(sizeof(dt_view_manager_t));
  dt_view_manager_init(darktable.view_manager);

  darktable.gui = (dt_gui_gtk_t *)malloc(sizeof(dt_gui_gtk_t));
  if(dt_gui_gtk_init(darktable.gui, argc, argv)) return 1;

  dt_control_load_config(darktable.control);
  strncpy(darktable.control->global_settings.dbname, filename, 512); // overwrite if relocated.

  darktable.imageio = (dt_imageio_t *)malloc(sizeof(dt_imageio_t));
  dt_imageio_init(darktable.imageio);

  int id = 0;
  if(image_to_load)
  {
    id = dt_image_import(1, image_to_load);
    if(id)
    {
      dt_film_open(1);
      DT_CTL_SET_GLOBAL(lib_image_mouse_over_id, id);
      dt_ctl_switch_mode_to(DT_DEVELOP);
    }
    else
    {
      // TODO: gtk window!
      fprintf(stderr, "[dt_init] could not open image file `%s'!\n", image_to_load);
    }
  }
  if(!id)
  {
    dt_ctl_switch_mode_to(DT_LIBRARY);
  }

  return 0;
}

void dt_cleanup()
{
  dt_ctl_switch_mode_to(DT_MODE_NONE);

  dt_control_write_config(darktable.control);
  dt_control_shutdown(darktable.control);

  dt_lib_cleanup(darktable.lib);
  free(darktable.lib);
  dt_view_manager_cleanup(darktable.view_manager);
  free(darktable.view_manager);
  dt_imageio_cleanup(darktable.imageio);
  free(darktable.imageio);
  dt_gui_gtk_cleanup(darktable.gui);
  free(darktable.gui);
  dt_image_cache_cleanup(darktable.image_cache);
  free(darktable.image_cache);
  dt_mipmap_cache_cleanup(darktable.mipmap_cache);
  free(darktable.mipmap_cache);
  dt_control_cleanup(darktable.control);
  free(darktable.control);
  dt_conf_cleanup(darktable.conf);
  free(darktable.conf);
  dt_points_cleanup(darktable.points);
  free(darktable.points);

  dt_camctl_destroy(darktable.camctl);
  dt_pwstorage_destroy(darktable.pwstorage);
  dt_fswatch_destroy(darktable.fswatch);

  sqlite3_close(darktable.db);
  pthread_mutex_destroy(&(darktable.db_insert));
  pthread_mutex_destroy(&(darktable.plugin_threadsafe));

  dt_exif_cleanup();
#ifdef HAVE_GEGL
  gegl_exit();
#endif
}

void dt_print(dt_debug_thread_t thread, const char *msg, ...)
{
  if(darktable.unmuted & thread)
  {
    va_list ap;
    va_start(ap, msg);
    vprintf(msg, ap);
    va_end(ap);
  }
}

void dt_gettime_t(char *datetime, time_t t)
{
  struct tm tt;
  (void)localtime_r(&t, &tt);
  strftime(datetime, 20, "%Y:%m:%d %H:%M:%S", &tt);
}

void dt_gettime(char *datetime)
{
  dt_gettime_t(datetime, time(NULL));
}

void *dt_alloc_align(size_t alignment, size_t size)
{
#if defined(__MACH__) || defined(__APPLE__) || (defined(__FreeBSD_version) && __FreeBSD_version < 700013)
  return malloc(size);
#else
  void *ptr = NULL;
  if(posix_memalign(&ptr, alignment, size)) return NULL;
  return ptr;
#endif
}

void 
dt_get_user_config_dir (char *data, size_t bufsize)
{
  g_snprintf (data,bufsize,"%s/.config/darktable",getenv("HOME"));  
  if (g_file_test (data,G_FILE_TEST_EXISTS)==FALSE) 
    g_mkdir_with_parents (data,0700);
}

void 
dt_get_user_cache_dir (char *data, size_t bufsize)
{
  g_snprintf (data,bufsize,"%s/.cache/darktable",getenv("HOME"));  
  if (g_file_test (data,G_FILE_TEST_EXISTS)==FALSE) 
    g_mkdir_with_parents (data,0700);
}


void 
dt_get_user_local_dir (char *data, size_t bufsize)
{
  g_snprintf(data,bufsize,"%s/.local",getenv("HOME"));
  if (g_file_test (data,G_FILE_TEST_EXISTS)==FALSE) 
    g_mkdir_with_parents (data,0700);

}

void dt_get_plugindir(char *datadir, size_t bufsize)
{
#if defined(__MACH__) || defined(__APPLE__)
  gchar *curr = g_get_current_dir();
  int contains = 0; for(int k=0;darktable.progname[k] != 0;k++) if(darktable.progname[k] == '/') { contains = 1; break; }
  if(darktable.progname[0] == '/') // absolute path
    snprintf(datadir, bufsize, "%s", darktable.progname);
  else if(contains) // relative path
    snprintf(datadir, bufsize, "%s/%s", curr, darktable.progname);
  else
  { // no idea where we have been called. use compiled in path
    g_free(curr);
    snprintf(datadir, bufsize, "%s/darktable", LIBDIR);
    return;
  }
  size_t len = MIN(strlen(datadir), bufsize);
  char *t = datadir + len; // strip off bin/darktable
  for(;t>datadir && *t!='/';t--); t--;
  if(*t == '.' && *(t-1) != '.') { for(;t>datadir && *t!='/';t--); t--; }
  for(;t>datadir && *t!='/';t--);
  strcpy(t, "/lib/darktable");
  g_free(curr);
#else
  snprintf(datadir, bufsize, "%s/darktable", LIBDIR);
#endif
}

void dt_get_datadir(char *datadir, size_t bufsize)
{
#if defined(__MACH__) || defined(__APPLE__)
  gchar *curr = g_get_current_dir();
  int contains = 0; for(int k=0;darktable.progname[k] != 0;k++) if(darktable.progname[k] == '/') { contains = 1; break; }
  if(darktable.progname[0] == '/') // absolute path
    snprintf(datadir, bufsize, "%s", darktable.progname);
  else if(contains) // relative path
    snprintf(datadir, bufsize, "%s/%s", curr, darktable.progname);
  else
  { // no idea where we have been called. use compiled in path
    g_free(curr);
    snprintf(datadir, bufsize, "%s", DATADIR);
    return;
  }
  size_t len = MIN(strlen(datadir), bufsize);
  char *t = datadir + len; // strip off bin/darktable
  for(;t>datadir && *t!='/';t--); t--;
  if(*t == '.' && *(t-1) != '.') { for(;t>datadir && *t!='/';t--); t--; }
  for(;t>datadir && *t!='/';t--);
  strcpy(t, "/share/darktable");
  g_free(curr);
#else
  snprintf(datadir, bufsize, "%s", DATADIR);
#endif
}

